sdk/communication/azure-communication-phonenumbers/src/main/java/com/azure/communication/phonenumbers/PhoneNumbersAsyncClient.java

```diff
@@ -129,12 +138,90 @@
-    public PollerFlux<PollResult<PhoneNumberSearchResult>, PhoneNumberSearchResult> beginSearchAvailablePhoneNumbers(
-        String countryCode, PhoneNumberSearchRequest searchRequest) {
-        return null;
+    public PollerFlux<PhoneNumberOperationResult, PhoneNumberSearchResult> beginSearchAvailablePhoneNumbers(
+        String countryCode, PhoneNumberSearchRequest searchRequest, Duration pollInterval) {
+        Objects.requireNonNull(countryCode, "'countryCode' cannot be null.");
+        Objects.requireNonNull(searchRequest, "'searchRequest' cannot be null.");
+        
+        if (pollInterval == null) {
+            pollInterval = defaultPollInterval;
+        }
+
+        return new PollerFlux<>(pollInterval,
+            searchAvailableNumbersInitOperation(countryCode, searchRequest),
+            searchAvailableNumbersPollOperation(),
+            cancelOperation(),
+            fetchFinalResultOperation());
+    }
+
+    private Function<PollingContext<PhoneNumberOperationResult>, Mono<PhoneNumberOperationResult>>
+        searchAvailableNumbersInitOperation(String countryCode, PhoneNumberSearchRequest searchRequest) {
+        return (pollingContext) -> {
+            return client.searchAvailablePhoneNumbersWithResponseAsync(countryCode, searchRequest)
+                .flatMap((PhoneNumbersSearchAvailablePhoneNumbersResponse response) -> {
+                    pollingContext.setData("operationId", parseIdFromUrl(response.getDeserializedHeaders().getOperationLocation()));
+                    pollingContext.setData("searchId", parseIdFromUrl(response.getDeserializedHeaders().getLocation()));
+                    return client.getOperationAsync(pollingContext.getData("operationId"))
+                    .flatMap((PhoneNumberOperationResult result) -> {
+                        return Mono.just(result);
+                    });
+                });
+        };
+    }
+
+    private String parseIdFromUrl(String url) {
+        String[] items = url.split("/");
+        String id = items[items.length - 1];
+        return id.substring(0, id.indexOf("?"));
+    }
+
+    private Function<PollingContext<PhoneNumberOperationResult>, Mono<PollResponse<PhoneNumberOperationResult>>>
+        searchAvailableNumbersPollOperation() {
+        return (pollingContext) -> { 
+            return client.getOperationAsync(pollingContext.getData("operationId"))
+            .flatMap(operation -> {
+                if (operation.getStatus().toString().equalsIgnoreCase(PhoneNumberOperationStatus.SUCCEEDED.toString())) {
+                    return Mono.just(new PollResponse<>(
+                        LongRunningOperationStatus.SUCCESSFULLY_COMPLETED, operation));
+                } else if (operation.getStatus().toString().equalsIgnoreCase(PhoneNumberOperationStatus.FAILED.toString())) {
+                    return Mono.just(new PollResponse<>(
+                        LongRunningOperationStatus.FAILED, operation));
```

Comment:
I think we would need to use `defer`, if we had passed `operation.getValue()` here but by passing the operation itself we might be fine.